torch_geometric/data/graph_store.py

```diff
@@ -28,14 +38,22 @@ class EdgeAttr(CastMixin):
     # meaningful for COO (CSC and CSR are sorted by definition)
     is_sorted: bool = False
 
-    # TODO support num_nodes here, default None, so users can specify this
-    # instead of relying on default inferral
+    # The number of nodes in this edge type. If set to None, will attempt to
+    # infer with the simple heuristic int(self.edge_index.max()) + 1
+    size: Optional[Tuple[int, int]] = None
 
-    def __init__(self, edge_type: Optional[Any], layout: EdgeLayout,
-                 is_sorted: bool = False):
+    # NOTE we define __init__ to force-cast layout
+    def __init__(
+        self,
+        edge_type: Optional[Any],
+        layout: EdgeLayout,
+        is_sorted: bool = False,
+        size: Optional[Tuple[int, int]] = None,
```

Comment:
Would it be better if we use `sparse_sizes` as `SparseTensor` did?

Reply:
They have the same meaning, but `sparse_sizes` wouldn't make sense for a general tensor.